File: frontend-v2/src/lib/ws/protocol.ts
// ═══════════════════════════════════════════════════════════════════
// Wire protocol types — TypeScript mirror of the generic WS envelope.
//
// Every WS message (both directions) is a JSON text frame using this
// shape. The `kind` field discriminates four lanes: control, notify,
// stream, error. See protocol.md for the full spec.
// ═══════════════════════════════════════════════════════════════════

// ---------------------------------------------------------------------------
// Envelope
// ---------------------------------------------------------------------------

/** Lane discriminator — determines which handler receives the message. */
export type EnvelopeKind = "control" | "notify" | "stream" | "error"

/** Resource target for messages scoped to a specific entity. */
export interface EnvelopeResource {
  type: string
  id: string
}

/**
 * Generic wire envelope. Every WS message uses this shape.
 *
 * Only `kind` and `op` are always present. Other fields are
 * lane-specific — see per-lane constants below.
 */
export interface Envelope {
  kind: EnvelopeKind
  op: string
  resource?: EnvelopeResource
  subId?: string
  seq?: number
  epoch?: string
  payload?: Record<string, unknown>
}

// ---------------------------------------------------------------------------
// Control lane ops
// ---------------------------------------------------------------------------

/** Client → server control ops. */
export const CONTROL_OP = {
  AUTH: "auth",
  PONG: "pong",
  SUBSCRIBE: "subscribe",
  UNSUBSCRIBE: "unsubscribe",
} as const

/** Server → client control ops. */
export const CONTROL_RESPONSE_OP = {
  CONNECTED: "connected",
  PING: "ping",
  SUBSCRIBED: "subscribed",
  UNSUBSCRIBED: "unsubscribed",
} as const

// ---------------------------------------------------------------------------
// Notify lane ops
// ---------------------------------------------------------------------------

export const NOTIFY_OP = {
  INVALIDATE: "invalidate",
} as const

// ---------------------------------------------------------------------------
// Stream lane ops
// ---------------------------------------------------------------------------

/** Server → client stream ops. */
export const STREAM_OP = {
  EVENT: "event",
  ENDED: "ended",
  GAP: "gap",
} as const

/** Client → server stream ops. */
export const STREAM_CLIENT_OP = {
  MESSAGE: "message",
} as const

// ---------------------------------------------------------------------------
// Error lane ops
// ---------------------------------------------------------------------------

export const ERROR_OP = {
  ERROR: "error",
} as const

/** Well-known error codes from the server. */
export const ERROR_CODE = {
  SUBSCRIBE_FAILED: "SUBSCRIBE_FAILED",
  RATE_LIMITED: "RATE_LIMITED",
  AUTH_FAILED: "AUTH_FAILED",
  INVALID_MESSAGE: "INVALID_MESSAGE",
} as const

// ---------------------------------------------------------------------------
// Notify event names (payload.event values)
// ---------------------------------------------------------------------------

export const NOTIFY_EVENT = {
  COMPLETED: "completed",
  ERROR: "error",
  CANCELLED: "cancelled",
  SPAWN_STARTED: "spawn_started",
  UPDATED: "updated",
  CREATED: "created",
} as const

// ---------------------------------------------------------------------------
// Resource types used in notify/stream messages
// ---------------------------------------------------------------------------

export const RESOURCE_TYPE = {
  TURN: "turn",
  THREAD: "thread",
  DOCUMENT: "document",
  PROPOSAL: "proposal",
} as const

// ---------------------------------------------------------------------------
// Connection state machine
// ---------------------------------------------------------------------------

/**
 * WS connection state machine.
 *
 * Transitions:
 *   disconnected → connecting → connected
 *   connected → reconnecting (on close/error)
 *   reconnecting → connecting (on backoff timer)
 *   any → disconnected (on explicit disconnect)
 */
export type ConnectionState =
  | "disconnected"
  | "connecting"
  | "connected"
  | "reconnecting"

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/**
 * Parse a raw WS text frame into an Envelope. Returns null if the
 * frame is not valid JSON or missing the required `kind`/`op` fields.
 */
export function parseEnvelope(raw: string): Envelope | null {
  try {
    const parsed: unknown = JSON.parse(raw)
    if (
      parsed == null ||
      typeof parsed !== "object" ||
      !("kind" in parsed) ||
      !("op" in parsed) ||
      typeof (parsed as Envelope).kind !== "string" ||
      typeof (parsed as Envelope).op !== "string"
    ) {
      return null
    }
    return parsed as Envelope
  } catch {
    return null
  }
}

/**
 * Build a control envelope. Convenience for client → server messages.
 */
export function controlEnvelope(
  op: string,
  payload?: Record<string, unknown>,
): Envelope {
  return { kind: "control", op, payload }
}

File: backend/internal/wsutil/protocol.go
package wsutil

import (
	"encoding/json"
	"errors"
	"fmt"
	"strings"
)

const (
	KindControl = "control"
	KindNotify  = "notify"
	KindStream  = "stream"
	KindError   = "error"
)

const (
	OpAuth         = "auth"
	OpConnected    = "connected"
	OpPing         = "ping"
	OpPong         = "pong"
	OpSubscribe    = "subscribe"
	OpSubscribed   = "subscribed"
	OpUnsubscribe  = "unsubscribe"
	OpUnsubscribed = "unsubscribed"
	OpEvent        = "event"
	OpEnded        = "ended"
	OpGap          = "gap"
	OpMessage      = "message"
	OpInvalidate   = "invalidate"
	OpError        = "error"
)

const (
	CodeSubscribeFailed = "SUBSCRIBE_FAILED"
	CodeRateLimited     = "RATE_LIMITED"
	CodeAuthFailed      = "AUTH_FAILED"
	CodeInvalidMessage  = "INVALID_MESSAGE"
	CodeNotSupported    = "NOT_SUPPORTED"
)

// Envelope is the generic wire message.
type Envelope struct {
	Kind     string          `json:"kind"`
	Op       string          `json:"op"`
	Resource *Resource       `json:"resource,omitempty"`
	SubId    string          `json:"subId,omitempty"`
	Seq      int64           `json:"seq,omitempty"`
	Epoch    string          `json:"epoch,omitempty"`
	Payload  json.RawMessage `json:"payload,omitempty"`
}

// Resource identifies a target resource.
type Resource struct {
	Type string `json:"type"`
	Id   string `json:"id"`
}

// ErrorPayload is sent inside error envelopes.
type ErrorPayload struct {
	Code    string `json:"code"`
	Message string `json:"message"`
}

var validKinds = map[string]struct{}{
	KindControl: {},
	KindNotify:  {},
	KindStream:  {},
	KindError:   {},
}

// Validate checks envelope shape required by the generic protocol.
func (e *Envelope) Validate() error {
	if e == nil {
		return errors.New("envelope is nil")
	}

	kind := strings.TrimSpace(e.Kind)
	if kind == "" {
		return errors.New("kind is required")
	}
	if _, ok := validKinds[kind]; !ok {
		return fmt.Errorf("unknown kind %q", e.Kind)
	}
	if strings.TrimSpace(e.Op) == "" {
		return errors.New("op is required")
	}

	return nil
}

// ParseEnvelope decodes and validates a JSON envelope.
func ParseEnvelope(data []byte) (*Envelope, error) {
	var env Envelope
	if err := json.Unmarshal(data, &env); err != nil {
		return nil, fmt.Errorf("unmarshal envelope: %w", err)
	}
	if err := env.Validate(); err != nil {
		return nil, err
	}
	return &env, nil
}

// NewErrorEnvelope returns a connection-scoped error envelope.
func NewErrorEnvelope(code, message string) Envelope {
	payload, _ := json.Marshal(ErrorPayload{Code: code, Message: message})
	return Envelope{
		Kind:    KindError,
		Op:      OpError,
		Payload: payload,
	}
}

// NewSubErrorEnvelope returns an error envelope for a specific subscription/resource.
func NewSubErrorEnvelope(subID string, resource *Resource, code, message string) Envelope {
	env := NewErrorEnvelope(code, message)
	env.SubId = subID
	env.Resource = resource
	return env
}

// ParseErrorPayload unmarshals the common error payload.
func ParseErrorPayload(raw json.RawMessage) (ErrorPayload, error) {
	var p ErrorPayload
	if len(raw) == 0 {
		return p, errors.New("empty error payload")
	}
	if err := json.Unmarshal(raw, &p); err != nil {
		return p, err
	}
	return p, nil
}
